text/0000-macro-hygiene-optout.md

```diff
@@ -0,0 +1,229 @@
+
+```rust
+#![feature(decl_macro)]
+#![feature(macro_hygiene_optout)]
+
+macro m($mod_name:ident) {
+    pub mod $mod_name {
+        pub const BAR: u32 = 123;
+    }
+}
+
+fn main() {
+    m!(foo);
+    let _ = foo::BAR;
+    //~^ ERROR cannot find value `BAR` in module `foo`
+}
+```
+
+## Guide: Example D
+[guide-example-d]: #guide-example-d
+
+```rust
+#![feature(decl_macro)]
+#![feature(macro_hygiene_optout)]
+
+macro m() {
+    pub mod #foo {
+        pub const BAR: u32 = 123;
+    }
+}
+
+fn main() {
+    m!();
+    let _ = foo::BAR;
+    //~^ ERROR cannot find value `BAR` in module `foo`
+}
+```
+
+## Meta-variables
+[meta-variables]: #meta-variables
+
+Hygiene escaping of meta-variables (i.e. `#$foo` and `$#foo`) does not have immediately obvious semantics or usefulness, so is explicitly disallowed for the present, and yields error messages.
+
+## Usage Notes
+[usage-notes]: #usage-notes
+
+While the motivation of this feature stems from defining or "exporting" new identifiers from macros to their call-site, where it is appropriate for the macro itself to choose/compute the name, it is clear from the above semantics that this feature allows for other potential uses cases. Most notably, one can use or "import" an identifier from their call-site. This, however, is *not* recommended, since this purpose is already fulfilled well by macro parameters. On the other hand, it is not explicitly disallowed, for two reasons:
+
+- Defining an identifier with call-site hygiene within that macro and then using it is a perfectly reasonable scenario.
+- Macro expansion is performed at the syntactical (token stream) level, before parsing, so definitions and uses cannot be easily distinguished.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The macro parser routine first parses the macro definition into a token stream (as before), but now also tags tokens and meta-variables with an enum value representing the kind of hygiene (definition-site or call-site). This is only enabled for new-style `macro!` macros (i.e. *decl_macro* or macros 2.0); for `macro_rules!` macros, the call-site sigil `#` is not handled specially, and gives rise to an error. The sigil is always treated as a separate token outside of macros, on the LHS of macro rules, and when not followed by an identifier on the RHS.
+
+It should be noted that the sigil `#` has nothing to do with the syntax for [raw identifiers][rfc-raw-identifiers], and can be disambiguated without problems. In fact, they can be used together without any issues, e.g. `#r#foo`.
+
+When the macro is invoked (expanded), each token tree is transcribed according to the following rules, depending on its hygiene tag.
+
+- *definition-site*: a normal mark is applied for the current expansion
+- *call-site*: a transparent mark is applied for the current expansion and the syntax context for every identifier in the token tree is changed to the syntax context of the call site.
```

Comment:
>and the syntax context for every identifier in the token tree is changed to the syntax context of the call site

What is this part about?
When a macro is expanded, an identifier gets an opaque mark added by default (`Span::def_site()` in proc macro API) or transparent mark if opt-out is in place (`Span::call_site()` in proc macro API), that's all what happens.

Reply:
Ah, I was slightly confused about how your transparent mark worked. I'll clarify that.

Reply:
Let me know if it's better now.